homeassistant/components/amcrest/__init__.py

```diff
@@ -19,193 +19,73 @@
-            if was_online:
-                self._start_recovery()
-            raise
-        except AmcrestError:
-            with self._wrap_lock:
-                was_online = self.available
-                errs = self._wrap_errors = self._wrap_errors + 1
-                offline = not self.available
-            _LOGGER.debug("%s camera errs: %i", self._wrap_name, errs)
-            if was_online and offline:
-                _LOGGER.error("%s camera offline: Too many errors", self._wrap_name)
-                self._start_recovery()
-            raise
-        with self._wrap_lock:
-            was_offline = not self.available
-            self._wrap_errors = 0
-            self._wrap_login_err = False
-        if was_offline:
-            self._unsub_recheck()
-            self._unsub_recheck = None
-            _LOGGER.error("%s camera back online", self._wrap_name)
-            self._wrap_event_flag.set()
-            dispatcher_send(self._hass, service_signal(SERVICE_UPDATE, self._wrap_name))
-        return ret
-
-    def _wrap_test_online(self, now):
-        """Test if camera is back online."""
-        _LOGGER.debug("Testing if %s back online", self._wrap_name)
-        try:
-            self.current_time
-        except AmcrestError:
-            pass
-
 
 def _monitor_events(hass, name, api, event_codes):
-    event_codes = ",".join(event_codes)
+    if "All" in event_codes["events"]:
+        event_codes_list = "All"
+    else:
+        event_codes_list = ",".join(
+            list(dict.fromkeys(event_codes["binary_sensors"] + event_codes["events"]))
+        )
     while True:
         api.available_flag.wait()
         try:
-            for code, start in api.event_actions(event_codes, retries=5):
-                signal = service_signal(SERVICE_EVENT, name, code)
-                _LOGGER.debug("Sending signal: '%s': %s", signal, start)
-                dispatcher_send(hass, signal, start)
+            for code, payload in api.event_actions(event_codes_list, retries=5):
+                _LOGGER.debug("Captured event: %s -- %s", code, payload)
+                if code in event_codes["binary_sensors"]:
+                    # Management of the different actions for binary sensors
+                    start = bool(
+                        payload["action"] == "Start" or payload["action"] == "Pulse"
+                    )
+                    signal = service_signal(SERVICE_EVENT, name, code)
+                    _LOGGER.debug("Sending signal: '%s': %s", signal, start)
+                    dispatcher_send(hass, signal, start)
+
+                if code in event_codes["events"] or "All" in event_codes["events"]:
+                    _LOGGER.debug(
+                        "Sending event to bus, event name: %s, payload: %s",
+                        code,
+                        payload,
+                    )
+                    hass.bus.fire(DOMAIN + "." + code, payload)
```

Comment:
I think you should add the camera name in the data of the event, or else it might be difficult for the user or automations to distinguish between different devices (some events do not get enough info sent by the device itself).

Reply:
Also, some devices send codes with very strange syntax. You might need to sanitize them since you make them part of the event name (in a much lower-scale PR I wrote I used only the domain as the event name to avoid such problems).

Reply:
i agree!
i have added device name and device serial numer to the payload

Can you give me an example of events with strange code

Reply:
My doorbell gives the events `CallNoAnswered`, then `SIPInvite` when pressing the button, and after some time gives a strange event with what looks like a truncated message, yielding no data member at all. Its name is `_CallNoAnswer_`

Reply:
yes I know the problem.
Different devices generate events with different names (my doorbell generates an event with the name "Invite" when I press the button)
I believe it is due to the fact that it is also possible to configure the device in different ways. My doorbell could be used:
- with an internal monitor
- via sip
- via your own app
So depending on how it has been configured maybe it generates different events


however _CallNoAnswer_ or CallNoAnswer do not think we have problems as the name of the events. 

you will use
amcres._CallNoAnswer_

Reply:
FYI - text wrapped with underscores is markup for italics. When referring to events such as **\_CallNoAnswer\_** you need to escape each underscore with a backslash so it appears properly without being italicized.

Are mixed case HA entities with or without underscores even valid?    Might be better to **lc** all event codes and ignore mixed case.

Please show an example of how to obtain the "Action" payload in event.data.   Specifically, on the AD110 doorbell, the event code is **\_DoTalkAction\_** with the Action **"Invite"** to signal a doorbell press.   This is quite different than how the DB11 doorbell does it.

Reply:
Can you indicate to me where in the home assistant documentation, it is indicated that the underscores represent the markup for italics?

Reply:
It's just markup language used for Github posts and many forums...  see:  https://guides.github.com/features/mastering-markdown/

Reply:
oh yes in github yes
but home assistant i don't think has problems with event names with underscores too

Reply:
Underscores in HA entity names are fine, I just have never seen a name start with an underscore before, and mixed case in the name (as your examples show) I believe is ignored and not an error.

Repeating myself (sorry), but I still would appreciate a complete example of how an AD110 doorbell press can be detected.

Here's the raw event code with payload when doorbell is pressed and answered...
```
Code=_DoTalkAction_;action=Pulse;index=0;data={\n   "Action" : "Invite",\n   "CallID" : "20200817204738@xxx@xxx",\n   "CallSrcMask" : 4\n'
Code=_DoTalkAction_;action=Pulse;index=0;data={\n   "Action" : "Answer",\n   "CallID" : "20200817204738@xxx@xxx",\n   "CallSrcMask" : 4,\n   "HangupReason" : "AnswerByPhone"\n'
Code=_DoTalkAction_;action=Pulse;index=0;data={\n   "Action" : "Hangup",\n   "CallID" : "20200817204738@xxx@xxx",\n   "CallSrcMask" : 4,\n   "HangupReason" : "HangupByPhone"\n
```

In the new Amcrest config UI, I assume you would manually enter into an event field "\_DoTalkAction\_"    Or is case ignored, and just "dotalkaction" without underscores suffice for a pattern match?   Or can the pattern match extend to the entire payload, not just the event code, and we can define the action "Invite" as the desired trigger?

If the new code can pattern match on ANY unique string in the payload, then we can simplify the doorbell press check a great deal and not have to have a condition template.  In this way, the AD110 test would be identical to the DB11 "Invite" test.

So this:
```
  trigger:
    platform: event
    event_type: amcrest.invite
```
versus:
```
  trigger:
    platform: event
    event_type: amcrest._dotalkaction_
  condition:
     - condition: template   
       value_template:  <some sort of event_data payload check for "Invite">
```

Reply:
That's not how it works. The event name is chosen at event creation, and thereafter checked against by listeners. There is no callback that every trigger runs so that the event type is a sort of regex against the body of the event.

**But** the event platform for triggers accepts an `event_data` argument that enables matching with partial content. See https://www.home-assistant.io/docs/automation/trigger/#event-trigger
Here you could probably use `action: Invite` in the  `event_data` section.

Reply:
Correct
you have to use
``
trigger:
    platform: event
    event_type: amcrest._dotalkaction_
    event_data:
        Action: Invite
``

PS
my doorbell does not insert the Action key inside the event payload

Reply:
Thanks to you both.   What I was suggesting would require a new feature in python-amcrest to use the "All" event that would be able to parse for any desired Action or a combination of 2 strings (Code AND Action) to pattern match.   So I'll drop that idea as it's certainly not applicable to this integration.

The doorbell press payload for the AD110 contains _**"Action" : "Invite"**_     It's unclear to me how the event_data becomes "**Action:  Invite**" where the quotes are ignored and spaces are removed.    But I'll take your word for it and will certainly test this.    Is it in a shape now that I can manually pull and test?